pkg/cli/image/manifest/source.go

```diff
@@ -0,0 +1,134 @@
+
+	"github.com/spf13/pflag"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	kcmdutil "k8s.io/kubectl/pkg/cmd/util"
+
+	operatorv1alpha1 "github.com/openshift/api/operator/v1alpha1"
+	operatorv1alpha1scheme "github.com/openshift/client-go/operator/clientset/versioned/scheme"
+	operatorv1alpha1client "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1alpha1"
+	"github.com/openshift/library-go/pkg/image/reference"
+)
+
+// ImageSourceOptions sets correct image source for oc adm commands
+type ImageSourceOptions struct {
+	ImageContentSourcePolicyFile string
+	ImageContentSourcePolicyList []operatorv1alpha1.ImageContentSourcePolicy
+}
+
+// Bind adds the options to the flag set.
+func (o *ImageSourceOptions) Bind(flags *pflag.FlagSet) {
+	flags.StringVar(&o.ImageContentSourcePolicyFile, "icsp-file", o.ImageContentSourcePolicyFile, "Path to an ImageContentSourcePolicy file.  If set, data from this file will be used to set source release image.")
+}
+
+// GetICSPs returns a list of ImageContentSourcePolicy objects found from file or from a connected cluster.
+func (o *ImageSourceOptions) GetICSPs(f kcmdutil.Factory) error {
+	// If ImageContentSourceFile is given, only add ImageContentSource from file, don't search cluster ICSP
+	if len(o.ImageContentSourcePolicyFile) != 0 {
+		icspData, err := ioutil.ReadFile(o.ImageContentSourcePolicyFile)
+		if err != nil {
+			return fmt.Errorf("unable to read ImageContentSourceFile %s: %v", o.ImageContentSourcePolicyFile, err)
+		}
+		if len(icspData) == 0 {
+			return fmt.Errorf("no data found in ImageContentSourceFile %s", o.ImageContentSourcePolicyFile)
+		}
+		icspObj, err := kruntime.Decode(operatorv1alpha1scheme.Codecs.UniversalDeserializer(), icspData)
+		if err != nil {
+			return fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", o.ImageContentSourcePolicyFile, err)
+		}
+		var icsp *operatorv1alpha1.ImageContentSourcePolicy
+		var ok bool
+		if icsp, ok = icspObj.(*operatorv1alpha1.ImageContentSourcePolicy); !ok {
+			return fmt.Errorf("could not decode ImageContentSourcePolicy from %s", o.ImageContentSourcePolicyFile)
+		}
+		o.ImageContentSourcePolicyList = append(o.ImageContentSourcePolicyList, *icsp)
+		return nil
+	}
+	restConfig, err := f.ToRESTConfig()
+	if err != nil {
+		// may or may not be connected to a cluster
+		// don't error if can't connect
+		klog.V(4).Infof("did not connect to an OpenShift 4.x server")
+		return nil
+	}
+	icspClient, err := operatorv1alpha1client.NewForConfig(restConfig)
+	if err != nil {
+		// may or may not be connected to a cluster
+		// don't error if can't connect
+		klog.V(4).Infof("did not connect to an OpenShift 4.x server")
+		return nil
+	}
+	icsps, err := icspClient.ImageContentSourcePolicies().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		// may or may not have access to ICSPs in cluster
+		// don't error if can't access ICSPs
```

Comment:
Hmm... I would want to know that I don't have access and I maybe should or that my mirror/copy/else will be missing something.

Reply:
updated the message there.